src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs

```diff
@@ -371,66 +371,113 @@ private static bool IsMultidimensionalInitializer(ImmutableArray<BoundExpression
+            avoidInPlace = false;
+            SpecialType specialElementType = SpecialType.None;
 
             if (!_module.SupportsPrivateImplClass)
             {
                 return false;
             }
 
-            var ctor = (MethodSymbol?)Binder.GetWellKnownTypeMember(this._module.Compilation, WellKnownMember.System_ReadOnlySpan_T__ctor_Pointer, _diagnostics, syntax: wrappedExpression.Syntax, isOptional: true);
-
-            if (ctor == null)
+            var rosPointerCtor = (MethodSymbol?)Binder.GetWellKnownTypeMember(_module.Compilation, WellKnownMember.System_ReadOnlySpan_T__ctor_Pointer, _diagnostics, syntax: wrappedExpression.Syntax, isOptional: true);
+            if (rosPointerCtor is null)
             {
                 return false;
             }
-
-            Debug.Assert(!ctor.HasUnsupportedMetadata);
-
+            Debug.Assert(!rosPointerCtor.HasUnsupportedMetadata);
+
+            // The purpose of this optimization is to replace a BoundArrayCreation with better code generation.
+            // We're looking for an expression like:
+            //     new ReadOnlySpan<T>(new T[] { const, const, ... })
+            //     new ReadOnlySpan<T>(new T[] { const, const, ... }, 0, length)
+            //     (ReadOnlySpan<T>)new T[] { const, const, ... }
+            // etc., and wrappedExpression is that array creation.  For single byte primitives, we can replace that
+            // with the equivalent of:
+            //     new ReadOnlySpan<T>((void*)PrivateImplementationDetails.DataField, Length)
+            // on all target platforms.  For primitives larger than a single byte, if the target platform exposes
+            // the RuntimeHelpers.CreateSpan method, we can emit it instead as:
+            //     RuntimeHelpers.CreateSpan(PrivateImplementationDetails.DataFieldToken)
+            // and for platforms that lack CreateSpan, as a span that wraps a lazily-initialized array:
+            //     new ReadOnlySpan<T>(PrivateImplementationDetails.ArrayField ??= new T[] { ... })
+            // For non-constant data, unsupported primitive types, and other variations, this optimization will fail
+            // and the method will return false indicating that no code was emitted.
+
+            ArrayTypeSymbol? arrayType = null;
             if (wrappedExpression is BoundArrayCreation ac)
             {
-                var arrayType = (ArrayTypeSymbol)ac.Type;
-                TypeSymbol elementType = arrayType.ElementType.EnumUnderlyingTypeOrSelf();
-
-                // NB: we cannot use this approach for element types larger than one byte
-                //     the issue is that metadata stores blobs in little-endian format
-                //     so anything that is larger than one byte will be incorrect on a big-endian machine
-                //     With additional runtime support it might be possible, but not yet.
-                //     See: https://github.com/dotnet/corefx/issues/26948 for more details
```

Comment:
Looking at this removed comment made me think that the endian-ness behavior should be mentioned in the `CreateSpan` doc. I assume that on certain platforms it will copy and rectify endian-ness, while on the more common platform no copy/rectification is needed.